tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -74,18 +75,21 @@ sub run_test {
         my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
 
         # Generate network load in guest
-        my $ping_url = "download.opensuse.org";
-        assert_script_run("ssh root\@$guest \"ping -f -c 500 $ping_url\"");
+        generate_vif_interrupts_in_guest($guest);
 
         # re-caculate the network interrupts distribution
         # the increased NIC IRQs should be distributed in CPU cores in balance
         my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
         my @increased_irqs_on_cpu;
         for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
             $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
-            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+            #at least a few interrupts on each cpu core
+            if ($increased_irqs_on_cpu[$cpu_id] < 10) {
+                # julie debug: will change 'record_soft_failure' to 'die' before pr is merge
+                record_soft_failure("IRQ are not balanced as the vif interrupts for CPU" . $cpu_id . " is " . $increased_irqs_on_cpu[$cpu_id]);
```

Comment:
3 guests are found IRQ are not balanced in the verfification test, http://openqa.qam.suse.cz/tests/37776. they are sle12sp4HVM, sle15HVM, sle15sp1PV. So here, do we abort test once a problematic guest is found here or continue test? @all @alice-suse

Reply:
I would suggest to continue test with record_soft_failure, qam reviewers always review every soft-failure every time.

Besides, regarding the irqbalance result on 12sp4/15/15sp1, we need to identify if this is product bug or this is expected behavior. They are older guests, and may have different behavior. Do they ever pass? What happens if you do not bind all interrupts to cpu0 and directly trigger network workload with irqbalance started? Are they balanced?

Reply:
> I would suggest to continue test with record_soft_failure, qam reviewers always review every soft-failure every time.

Ok, I'll keep the current soft_failure.

> Besides, regarding the irqbalance result on 12sp4/15/15sp1, we need to identify if this is product bug or this is expected behavior. They are older guests, and may have different behavior. Do they ever pass? What happens if you do not bind all interrupts to cpu0 and directly trigger network workload with irqbalance started? Are they balanced?

No, they have never passed. other newer guests are irq balanced with latest updated packages only.

if I would not bind all interrupts to cpu0, it most likely all IRQ still were assigned on CPU0, because there was zero interrupts on other CPUs initially. ie. no balance at all regardless what I did.

Reply:
Or because they are LTSS guests, we don't update them to latest LTSS repo?

Reply:
In MU regular test flow, the guests should have been updated to latest LTSS. But here in your validation job, I do not see its parent jobs in "Dependency". Do you skip the step? But this should be able to be validated by checking the guest inside repo and the irqbalance rpm version. I checked again the bug [1178477](https://bugzilla.suse.com/show_bug.cgi?id=1178477), it seems the fix was submitted to 15SP1 LTSS already, but not sure whether it has been RELEASED to customer's LTSS repo. If the guest's installed irqbalance rpm >=   irqbalance-1.4.0, it should be with the fix.

```
SUSE-RU-2021:1192-1: An update that has two recommended fixes can now be installed.

Category: recommended (important)
Bug References: 1178477,1183405
CVE References: 
JIRA References: 
Sources used:
SUSE Linux Enterprise Server 15-SP1-LTSS (src):    irqbalance-1.4.0-7.6.1
```

Reply:
> In MU regular test flow, the guests should have been updated to latest LTSS. But here in your validation job, I do not see its parent jobs in "Dependency". Do you skip the step? 

the parent job is listed in 'dependency', please check http://openqa.qam.suse.cz/tests/37776#dependencies

> But this should be able to be validated by checking the guest inside repo and the irqbalance rpm version. I checked again the bug 1178477, it seems the fix was submitted to 15SP1 LTSS already, but not sure whether it has been RELEASED to customer's LTSS repo. If the guest's installed irqbalance rpm >= irqbalance-1.4.0, it should be with the fix.

in 15sp1 guest, 
irqbalance-1.4.0-7.3.1.x86_64(http://openqa.qam.suse.cz/tests/37776#step/xen_guest_irqbalance/716)
so the fix has not gone into our guest yet.

Reply:
In the latest run http://openqa.qam.suse.cz/tests/38110, only sle15HVM failed due to IRQ assigned on CPU0 only, sle15sp1PV & sle12sp4HVM passed.
irqbalance-1.4.0-7.6.1.x86_64 in 15sp1 now.

Reply:
@Julie-CAO If this is final status that the fix for 15GA is not yet in update repo, in the final test code for MU,  will it be better to, at the beginning of this test, add a check for irqbalance version, if below the fix version, skip the test and  add the bug info in the record_soft_fail info?

Reply:
Hi @alice-suse , I did not find the package listed in the bug for SLE15. https://bugzilla.suse.com/show_bug.cgi?id=1178477. So does it mean no fix goes into SLE15 update repo?

Reply:
@Julie-CAO Likely, but better double confirm in the bug.

Reply:
there will be no fix in SLE15 as the Olaf said no customer requests in the bug. So I skip SLE15 guest in the new push, the verification run passed, http://openqa.qam.suse.cz/tests/38164.